pkg/package-server-manager/controller.go

```diff
@@ -0,0 +1,192 @@
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		if apierrors.IsNotFound(err) {
+			log.Info("csv does not exist attempting to create it")
+			csv, err = manifests.NewPackageServerCSV(r.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			_ = ensureCSV(log, csv, r.getHighlyAvailableMode())
+			if err := r.Client.Create(ctx, csv); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	if !csv.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("requeuing after encountering a delete event")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	orig := client.MergeFrom(csv.DeepCopy())
+	if !ensureCSV(log, csv, r.getHighlyAvailableMode()) {
+		log.Info("no updates to the csv are necessary")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Client.Patch(ctx, csv, orig); err != nil {
+		log.Error(err, "failed to patch the csv")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PackageServerCSVReconciler) infrastructureHandler(obj client.Object) []reconcile.Request {
+	log := r.Log.WithValues("infrastructure", obj.GetName())
+
+	if obj.GetName() != infrastructureName {
+		return nil
+	}
+
+	var infra configv1.Infrastructure
+	if err := r.Client.Get(context.Background(), types.NamespacedName{Name: obj.GetName()}, &infra); err != nil {
+		return []reconcile.Request{}
+	}
+
+	r.setTopologyMode(infra.Status.ControlPlaneTopology)
+	log.Info("requeueing the packageserver deployment after encountering infrastructure event")
+
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Name:      name,
+				Namespace: r.Namespace,
+			},
+		},
+	}
+}
+
+func (r *PackageServerCSVReconciler) setTopologyMode(topologyMode configv1.TopologyMode) {
+	r.Lock.Lock()
+	defer r.Lock.Unlock()
+
+	if topologyMode == configv1.SingleReplicaTopologyMode {
+		r.HighlyAvailableMode = false
+		return
+	}
+	r.HighlyAvailableMode = true
+}
+
+func (r *PackageServerCSVReconciler) getHighlyAvailableMode() bool {
+	r.Lock.Lock()
+	defer r.Lock.Unlock()
+
+	return r.HighlyAvailableMode
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PackageServerCSVReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	predicateNameFilter := predicate.NewPredicateFuncs(func(object client.Object) bool {
+		return object.GetName() == r.Name && object.GetNamespace() == r.Namespace
+	})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&olmv1alpha1.ClusterServiceVersion{}, builder.WithPredicates(predicateNameFilter)).
+		Watches(&source.Kind{Type: &configv1.Infrastructure{}}, handler.EnqueueRequestsFromMapFunc(r.infrastructureHandler)).
+		Complete(r)
+}
+
+func ensureCSV(logger logr.Logger, csv *olmv1alpha1.ClusterServiceVersion, highlyAvailableMode bool) bool {
+	var modified bool
+
+	deploymentSpecs := csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs
+	if len(deploymentSpecs) != 1 {
+		logger.Info("not processing csv with more than one nested deployment")
+		return false
```

Comment:
IIUC, this means that a user can remove (or add) a deployment spec to the CSV and the packageserver manager will do nothing to correct the CSV. This is a change in behavior from what happens with CVO today (since it will stomp those changes).

My question is, do we want to diverge from CVO's behavior?

Reply:
That was previously a stub that was used to avoid blindly indexing into that array of deploymentSpec(s). I'm wondering whether we'd prefer to check if we're handling a CSV that contains more nested deploymentSpec than what we're expecting, we should just serialize the YAML manifest directly, and toggle the modified bool to true. That way, even if there are no modifications needed on the replica/rolloutStrategy configuration side-of-things, the CSV parameter will be assigned the new deploymentSpec. I think we'd like to avoid exposing a latch for users/controllers/etc. to make changes to this CSV that we don't support.

Reply:
Updated to a helper function that validates whether the deployment and individual containers match the expectations we have and if not, the parent site overrides the current CSV with the default YAML manifest configuration that gets passed to the ensureCSV function.